docs/rfcs/1005-realm-key-rotation.md

```diff
@@ -0,0 +1,821 @@
+   - the maintenance is done, so the new key works. All is good.
+   - the maintenance is still ongoing, so the client has to look into the previous
+     versions of the user manifest to find the old key. This is messy :'(
+
+## 2 General approach
+
+The idea is as follow:
+
+- remove the whole maintenance logic: a realm is always accessible in read&write
+- remove the whole re-encryption logic: realm vlobs are never reencrypted
+- remove the whole message logic: we use a special ad-hoc command to access the
+  realm's keys (see below)
+- a realm now has it manifests encrypted with multiple keys, at any time a new
+  key can be added
+- each key for a given realm has an index
+- only the last key should be used to encrypt new manifest
+- all those keys are serialized together in a `RealmKeysBundle` which is encrypted by the
+  "keys-bundle key" symmetric key and stored in a dedicate table in the server database
+- the "keys-bundle key" is serialized in a `RealmKeysBundleAccess`. This document is signed
+  by the author of the key rotation and encrypted for each member of the realm with their
+  public key
+- for sharing a realm, the last `RealmKeysBundleAccess` is encrypted for the new user.
+  The `RealmKeysBundleAccess` also contains the workspace label.
+- keys-bundle & keys-bundle access are never stored on-disk client-side: they are instead
+  lazily fetched from server when needed and kept on RAM.
+- realm initial key is lazily created by doing the very first key rotation (typically during
+  minimal sync right after having created the realm).
+- key rotation also involves the creation of a `RealmKeyRotationCertificate` certificate to
+  allow auditability (knowing when and by whom the key rotation occurred)
+- key rotation certificate contains the index of the key along with it algorithm and a
+  canary (i.e. an encrypted payload to ensure the key obtained from the keys-bundle is genuin)
+- key rotation consists of creating a new `RealmKeysBundle` with one additional
+  key, then submit it (along with the `RealmKeyRotationCertificate` and the per-user
+  `RealmKeysBundleAccess`) to the server with the new `workspace_rotate_keys` API command.
+- given each key rotation is expected to add a single key to the bundle, the key index is
+  also used to identify the key rotation and the keys-bundle
+- vlob create/update/read API commands contains a field with the index of the key being used
+- user manifest no longer store the realm key, only the realm id and name. The realm keys
+  are only fetched when needed (i.e. in case of out/in sync) by using the new
+  `workspace_get_keys_bundle` API command.
+- `workspace_get_keys_bundle` allows to get previous keys-bundles if any, this is needed
+  to self-heal in case the new keys-bundle is corrupted.
+- in case the "keys-bundle key" becomes corrupted, the self-heal operation consist of
+  retrieving the last valid "keys-bundle key" and doing a new key rotation from there.
```

Comment:
> Unless the server is capable of detected the problem itself 

`KeyBundle` is encrypted so server has not power here !

> the last key bundle is requested and, IF a problem arises, the previous key bundles are requested and a new key rotation is performed
>
> Should we consider "tagging" the corrupted key bundle so it is no longer used and/or can be cleaned later?

You are correct: a corrupted key rotation could introduce a key that doesn't correspond to the key rotation certiifcate's canary. In this case the self-heal operation must include the key (given it index is the position in the keys array), but doesn't know what to include !

The solution is to sign the keybundle (so that we trust whatever key are in the key bundle, even if some are missing), and allow missing keys (providing the keys in the `KeyBundle` as `Vec<Option<Bytes>>`

Reply:
> The solution is to sign the keybundle (so that we trust whatever key are in the key bundle, even if some are missing), and allow missing keys (providing the keys in the KeyBundle as Vec<Option<Bytes>>

I've slightly changed the solution: It is much simpler to just consider a KeyBundle valid as long as it is correctly signed and contains the right number of keys. This way there is no need to handle a `None` value for the keys, and the self-heal operation becomes trivial: just recursively get the previous key bundle until one of them is valid.